File: apps/service-java/src/main/java/com/finflow/model/Expense.java
package com.finflow.model;

import lombok.*;
import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.mongodb.core.mapping.Document;
import org.springframework.data.mongodb.core.index.CompoundIndex;
import org.springframework.data.mongodb.core.index.Indexed;

import java.math.BigDecimal;
import java.time.Instant;
import java.util.List;
import java.util.Map;

/**
 * Expense document — shared MongoDB schema between Node.js and Java services.
 *
 * The Node-Flow Agent writes to the "expenses" collection.
 * The Java-Core Agent reads from it for audit/analysis.
 */
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Document(collection = "expenses")
@CompoundIndex(name = "user_category_date_idx", def = "{'userId': 1, 'category': 1, 'createdAt': -1}")
@CompoundIndex(name = "user_date_idx", def = "{'userId': 1, 'createdAt': -1}")
@CompoundIndex(name = "category_amount_idx", def = "{'category': 1, 'amount': -1}")
public class Expense {

    @Id
    private String id;

    @Indexed
    private String userId;

    private BigDecimal amount;

    @Indexed
    private String category;

    private String description;

    private String currency;

    private String source;

    private String processedBy;

    @Indexed
    private String status;

    private List<String> tags;

    private Map<String, Object> metadata;

    @CreatedDate
    private Instant createdAt;

    @LastModifiedDate
    private Instant updatedAt;
}

File: apps/service-node/src/models/Expense.ts
import mongoose, { Schema, Document } from 'mongoose';

export interface IExpense extends Document {
  userId: string;
  amount: number;
  category: string;
  description: string;
  currency: string;
  source: 'manual' | 'ai-agent' | 'import';
  processedBy: string;
  status: 'pending' | 'confirmed' | 'disputed' | 'deleted';
  tags: string[];
  metadata: Record<string, unknown>;
  createdAt: Date;
  updatedAt: Date;
}

const ExpenseSchema = new Schema<IExpense>(
  {
    userId: {
      type: String,
      required: true,
      index: true,
    },
    amount: {
      type: Number,
      required: true,
      min: 0,
    },
    category: {
      type: String,
      required: true,
      index: true,
    },
    description: {
      type: String,
      required: true,
      trim: true,
    },
    currency: {
      type: String,
      default: 'USD',
      uppercase: true,
      minlength: 3,
      maxlength: 3,
    },
    source: {
      type: String,
      enum: ['manual', 'ai-agent', 'import'],
      default: 'ai-agent',
    },
    processedBy: {
      type: String,
      default: 'node-flow-agent',
    },
    status: {
      type: String,
      enum: ['pending', 'confirmed', 'disputed', 'deleted'],
      default: 'confirmed',
      index: true,
    },
    tags: {
      type: [String],
      default: [],
    },
    metadata: {
      type: Schema.Types.Mixed,
      default: {},
    },
  },
  {
    timestamps: true,
    collection: 'expenses',
  },
);

// Compound indexes for efficient queries from both Node & Java services
ExpenseSchema.index({ userId: 1, category: 1, createdAt: -1 });
ExpenseSchema.index({ userId: 1, createdAt: -1 });
ExpenseSchema.index({ category: 1, amount: -1 });
ExpenseSchema.index({ status: 1, createdAt: -1 });

export const Expense = mongoose.model<IExpense>('Expense', ExpenseSchema);
